### ISSUE TITLE
2.2.6 "Special price date from" Failed to parse time string

### ISSUE THREAD

**Issue by Cristiano81** [magento/magento2#18158]
### Preconditions
1. Magento 2.2.6
2. Account locale set to it_IT

### Steps to reproduce
1. Magento 2.2.6 create a new product with a price
2. Edit the new product and add a special price

### Expected result
1. The product has been saved

### Actual result
1. DateTime::__construct(): Failed to parse time string (20/09/2018) at position 0 (2): Unexpected character

This worked on 2.2.5

**Comment by magento-engcom-team** [comment]
Hi @Cristiano81. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@Cristiano81 do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `G1 Passed` will be added to the issue automatically. Please, edit issue description if needed, until label `G1 Passed` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `G2 Passed` label to the issue by yourself.

- [x] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [x] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [x] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [x] 6. Add label `acknowledged` once verification is complete. 

- [x] 7. Make sure that automatic system confirms that report is acknowledged.

**Comment by ghost** [comment]
Hi @Cristiano81  thank you for you report, i'm not able to reproduce following steps you described. 
I'm have  saved product.

**Comment by magento-engcom-team** [comment]
@engcom-backlog-nazar Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-95124`, `MAGETWO-95125` were created

**Comment by Cristiano81** [comment]
yes i can reproduce on vanilla Magento https://i-18157-2-2-6.engcom.dev.magento.com
i have to add some precondition
1. Magento 2.2.6
2. Account locale set to it_IT
3. Locale options: locale italiano (italia) 

to reproduce  the issue after edit the product to insert the special price, save it again, elsewhere in first instance it works (when special price date from is empty)

**Comment by Cristiano81** [comment]
i think i have found the reason. 
in /vendor/magento/module-catalog/Controller/Adminhtml/Product/Initialization/Helper.php
has been added a new function on line 463
```
private function convertSpecialFromDateStringToObject($productData)
    {
        if (isset($productData['special_from_date']) && $productData['special_from_date'] != '') {
            $productData['special_from_date'] = new \DateTime($productData['special_from_date']);
        }

        return $productData;
    }
```

comment the call to that function on line 162
`//$productData = $this->convertSpecialFromDateStringToObject($productData);`

the product has been saved successfully again

**Comment by magento-engcom-team** [comment]
Hi @hiren0241. Thank you for working on this issue. 
Looks like this issue is already verified and confirmed. But if your want to validate it one more time, please, go though the following instruction:
- [ ] 1. Add/Edit `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.

- [ ] 2. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 3. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 4. If the issue is not relevant or is not reproducible any more, feel free to close it.

**Comment by sidolov** [comment]
Hi @Cristiano81. Thank you for your report. 
The issue has been fixed in magento/magento2#18578 by @hiren0241 in 2.3-develop branch
Related commit(s):
 - [4e382cc735a883c9dcd8eaeeedc185d7faa06a7f](https://github.com/magento/magento2/commit/4e382cc735a883c9dcd8eaeeedc185d7faa06a7f)
 - [fd8a5a90b32431fbdd275000624723b106c81841](https://github.com/magento/magento2/commit/fd8a5a90b32431fbdd275000624723b106c81841)
 - [88592facd5d34872ac2f2051dda246703091cacc](https://github.com/magento/magento2/commit/88592facd5d34872ac2f2051dda246703091cacc)

The fix will be available with the upcoming 2.3.1 release.

**Comment by paul-blundell** [comment]
@sidolov Will this also fix the From date being parsed as US date-format and the To date being parsed as the sites current locale.

E.g. Dates set to UK format e.g.  DD/MM/YYYY
Enter **From** date of 12/11/2018
Enter **To** date of 13/11/2018

I get an error saying the to date must be later than the from date.
If I switch the From date to 11/12/2018 it will save correctly and I can see in the data from an API call it is correctly returning the dates I wanted. (2018-11-12 and 2018-11-13)

Also, any idea when we can expect 2.3?

**Comment by paul-blundell** [comment]
This is a fairly major problem, why the decision to wait until 2.3.1 to rollout any fix?

**Comment by ghost** [comment]
@Cristiano81  this issue present in 2.3-develop instance with Account locale set to en_GB
![selection_197](https://user-images.githubusercontent.com/40756439/48264083-f7196580-e430-11e8-9562-9e7fafb49639.png)

**Comment by sidolov** [comment]
Hi @paul-blundell , we expecting 2.3 release very soon, but current fix will be available in 2.3.1 version according to the previous comment.

**Comment by paul-blundell** [comment]
@sidolov but that fix apparently hasn't resolved the issue anyway. Any idea when this will be resolved?

**Comment by alefemafra** [comment]
I created a fix to solve this problem, following the hint of @Cristiano81, follow the link of the repository, just download and put in app/code, and follow the installation standard of the module.

https://github.com/alefemafra/module-fix-date-product-error

**Comment by ghost** [comment]
Hi @Cristiano81 , thank you for you report, this issue has already fixed in 2.3-develop branch, and will be available on 2.3.1 release.

**Comment by magento-engcom-team** [comment]
Hi @Cristiano81. Thank you for your report. 
The issue has been fixed in magento/magento2#21273 by @tufahu in 2.2-develop branch
Related commit(s):
 - [4d9e6c8f946594af042c9d8eec8e82ecc63aa307](https://github.com/magento/magento2/commit/4d9e6c8f946594af042c9d8eec8e82ecc63aa307)
 - [523567461e74dd4580fe8e4e59d71aeae8191042](https://github.com/magento/magento2/commit/523567461e74dd4580fe8e4e59d71aeae8191042)
 - [ad5bf08df72a5115b2777ebf55b2fc8530c88d3e](https://github.com/magento/magento2/commit/ad5bf08df72a5115b2777ebf55b2fc8530c88d3e)

The fix will be available with the upcoming 2.2.9 release.
